### PR TITLE
Add `wallet.all_balances` endpoint

### LINKED ISSUE #477: Add paginations for wallet balances in Admin Panel

### PR DESCRIPTION
**Pull request by ripzery**
Issue/Task Number: #477 

Closes #477 
Requires #730

# Overview

This PR implements `wallet.all_balances` to allow querying paginated balances.

# Implementation Details

I'm so confuse with many files which I have to update, so I think I should do something that help the reviewer.

Here is the **roughly** drawing to track where does the logic go through. It is not so accurate, but I hope it helps  🙏

OMG-146

> A picture is worth a thousand words. So mostly is something like..

runn

> A baton is tokens which passing all the wayy

# Usage

call `/wallet.all_balances`

**Request**
```json
{
  "sort_by": "id",
  "sort_dir": "desc",
  "per_page": 1,
  "start_after": "tok_KNC_01d4m3wf46crk0v2y7arpyzskz",
  "start_by": "id",
  "address": "opxc251377416372"
}
```

**Response**
```json
{
  "version": "1",
  "success": true,
  "data": {
    "pagination": {
      "start_by": "id",
      "start_after": "tok_KNC_01d4m3wf46crk0v2y7arpyzskz",
      "per_page": 1,
      "is_last_page": false,
      "count": 1
    },
    "object": "list",
    "data": [
      {
        "token": {
          "updated_at": "2019-02-26T05:19:21.325039Z",
          "symbol": "ETH",
          "subunit_to_unit": 1000000000000000000,
          "object": "token",
          "name": "Ether",
          "metadata": {},
          "id": "tok_ETH_01d4m3wf7cn56wg939bmzkdp60",
          "encrypted_metadata": {},
          "enabled": true,
          "created_at": "2019-02-26T05:19:21.325039Z"
        },
        "object": "balance",
        "amount": 8e+21
      }
    ]
  }
}
```

# Impact

Deploy as usual.

### CODE REVIEW

**Comment by T-Dnzt** [comment]
I know this is not ready, but the endpoint should be`/wallet.get_balances` to follow our conventions.

**Comment by unnawut** [comment]
Ideally this also should reflect and be merged after #730.

- [ ] Wait for #730 to be merged
- [ ] Update this PR's permission checking to #730's way.

File: apps/admin_api/priv/swagger/balance/response_schemas.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+BalancesResponseSchema:
+  description: The response schema for a list of balance
```
Comment (unnawut) [inline apps/admin_api/priv/swagger/balance/response_schemas.yaml]:
```suggestion
  description: The response schema for a list of balances
```

File: apps/ewallet/lib/ewallet/web/balance_loader.ex [hunk 2]
```diff
@@ -29,6 +29,10 @@ defmodule EWallet.Web.BalanceLoader do
     BalanceFetcher.all(%{"wallet" => wallet})
   end
 
+  def add_balances(%Wallet{} = wallet, tokens) do
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/web/balance_loader.ex]: This needs to be covered by tests.

File: apps/local_ledger/lib/local_ledger/cached_balance.ex [hunk 3]
```diff
@@ -31,12 +31,14 @@ defmodule LocalLedger.CachedBalance do
...
+  def all(wallet, attrs \\ %{})
+
   @doc """
   Get all the balances for the given wallet or wallets.
   """
-  @spec all(%Wallet{} | [%Wallet{}]) :: {:ok, map()}
-  def all(wallet_or_wallets) do
-    {:ok, get_balances(wallet_or_wallets)}
+  def all(wallet_or_wallets, attrs) do
+    {:ok, get_balances(wallet_or_wallets, attrs)}
```
Comment (unnawut) [inline apps/local_ledger/lib/local_ledger/cached_balance.ex]: This change (the `attrs`) needs to be covered by tests.

File: apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex [hunk 4]
```diff
@@ -86,25 +94,38 @@ defmodule EWallet.BalanceFetcher do
...
 
-  defp query_and_add_balances(wallets) when is_list(wallets) do
+  defp query_and_add_balances(wallet_or_wallets, attrs \\ %{})
+
+  defp query_and_add_balances(wallets, _) when is_list(wallets) do
     wallets
     |> Enum.map(fn wallet -> wallet.address end)
     |> LocalLedger.Wallet.all_balances()
-    |> process_response(wallets, :all)
+    |> process_response(wallets, {:all})
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex]: Can remove `{:all}` param from here and also from the called function?

File: apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex [hunk 5]
```diff
@@ -86,25 +94,38 @@ defmodule EWallet.BalanceFetcher do
...
-    |> process_response(wallet, :all)
+    |> LocalLedger.Wallet.all_balances(attrs)
+    |> process_response(wallet, {:some, tokens})
+  end
+
+  defp query_and_add_balances(wallet, _) do
+    balances =
+      wallet.address
+      |> LocalLedger.Wallet.all_balances()
+      |> process_response(wallet, {:all})
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex]: Can refactor to `process_response(wallet, :all_tokens)`?

File: apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex [hunk 6]
```diff
@@ -86,25 +94,38 @@ defmodule EWallet.BalanceFetcher do
...
+    |> process_response(wallets, {:all})
   end
 
-  defp query_and_add_balances(wallet) do
+  defp query_and_add_balances(wallet, %{"tokens" => tokens} = attrs) do
     wallet.address
-    |> LocalLedger.Wallet.all_balances()
-    |> process_response(wallet, :all)
+    |> LocalLedger.Wallet.all_balances(attrs)
+    |> process_response(wallet, {:some, tokens})
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/fetchers/balance_fetcher.ex]: Can refactor to `process_response(wallet, tokens)`?

File: apps/admin_api/lib/admin_api/v1/controllers/balance_controller.ex [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    BalanceLoader,
+    V1.TokenOverlay
+  }
+
+  @doc """
+  Retrieves a paginated list of balances by given wallet address.
+  """
+  def all_for_wallet(conn, %{"address" => address} = attrs) do
+    with %Wallet{} = wallet <- Wallet.get(address) || {:error, :unauthorized},
+         {:ok, _} <- authorize(:get, conn.assigns, wallet),
```
[the remark this answers is not in the document]
Comment (ripzery) [reply apps/admin_api/lib/admin_api/v1/controllers/balance_controller.ex]: :bangbang:

File: apps/admin_api/priv/swagger/wallet/request_bodies.yaml [hunk 8]
```diff
@@ -44,6 +44,23 @@ WalletUserBody:
         example:
           id: usr_01cc02x0v98qcctvycfx4vsk8x
 
+BalanceAllForWalletBody:
+  description: The parameters to use for listing the balances for a wallet
```
Comment (mederic-p) [inline apps/admin_api/priv/swagger/wallet/request_bodies.yaml]: Can we specify that the pagination params are based on `token` to make it clear?

**Pull request merged**